# Problem Statement:

Atliq Grands, a leading player in India's hotel industry, is currently facing challenges with rising competition and declining revenue. The company is also experiencing a loss of market share. To combat these challenges, the management team has decided to harness data and strategic insights. With limited internal expertise in data analysis, they have sought external support to assess the situation, deliver actionable insights, and develop effective strategies for customer retention and revenue growth.

# Project Goal:

The goal of this project is to conduct a comprehensive analysis of the hotel industry domain using Python. By leveraging data-driven insights and strategic recommendations, the objective is to empower Atliq Grands to make informed decisions. The ultimate aim is to assist Atliq Grands in regaining a competitive edge in the market, addressing their challenges of revenue decline and market share loss effectively.

# Datasets:
* dim_date.csv
* dim_hotels.csv
* dim_rooms.csv
* fact_bookings.csv
* new_data_august.csv
* fact_aggregated_bookings.csv

### Importing Data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly as pt
import plotly.express as px
import warnings
warnings.filterwarnings("ignore")
import plotly.io as pio
pio.renderers.default = "notebook_connected"

C:\Users\Shridatta\anaconda3\lib\site-packages\scipy\__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5)
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "


In [2]:
df_date = pd.read_csv('datasets/dim_date.csv')
df_hotels = pd.read_csv('datasets/dim_hotels.csv')
df_rooms = pd.read_csv('datasets/dim_rooms.csv')
df_agg_bookings = pd.read_csv('datasets/fact_aggregated_bookings.csv')
df_bookings = pd.read_csv('datasets/fact_bookings.csv')

In [3]:
df_bookings.head()

booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   
3  May012216558RT14        16558   2022-04-28    2022-05-01    2022-05-02   
4  May012216558RT15        16558   2022-04-27    2022-05-01    2022-05-02   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   
3          2           RT1           others            NaN      Cancelled   
4          4           RT1    direct online            5.0    Checked Out   

   revenue_generated  revenue_realized  
0              10010             10010  
1               9100              3640  
2               9100              9100  
3               9100              3640  
4              10920             10920

In [4]:
df_bookings.shape

(134590, 12)

In [5]:
df_bookings.room_category.unique()

array(['RT1', 'RT2', 'RT3', 'RT4'], dtype=object)

In [6]:
df_bookings.booking_platform.unique()

array(['direct online', 'others', 'logtrip', 'tripster', 'makeyourtrip',
       'journey', 'direct offline'], dtype=object)

In [7]:
df_bookings.booking_platform.value_counts()

others            55066
makeyourtrip      26898
logtrip           14756
direct online     13379
tripster           9630
journey            8106
direct offline     6755
Name: booking_platform, dtype: int64

### Data Cleaning And EDA

In [8]:
# Count the number of bookings per platform
booking_counts = df_bookings['booking_platform'].value_counts().reset_index()
booking_counts.columns = ['Booking Platform', 'Number of Bookings']

# Create a bar chart
fig = px.bar(booking_counts, 
             x='Booking Platform', 
             y='Number of Bookings', 
             title='Number of Bookings per Platform',
             color='Number of Bookings',
             color_continuous_scale=px.colors.sequential.Blues)

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='Booking Platform',
    yaxis_title='Number of Bookings',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
     autosize=False,width=950,height=600
)

# Adding value annotations on top of bars
fig.for_each_trace(lambda t: t.update(text=[f"{v}" for v in t.y], textposition='outside'))

# Show the plot
fig.show()


In [9]:
df_bookings.describe()

property_id      no_guests  ratings_given  revenue_generated  \
count  134590.000000  134590.000000   56683.000000      134590.000000   
mean    18061.113493       2.036808       3.619004       14916.013188   
std      1093.055847       1.031766       1.235009        6452.868072   
min     16558.000000       1.000000       1.000000        6500.000000   
25%     17558.000000       1.000000       3.000000        9900.000000   
50%     17564.000000       2.000000       4.000000       13500.000000   
75%     18563.000000       2.000000       5.000000       18000.000000   
max     19563.000000       6.000000       5.000000       45220.000000   

       revenue_realized  
count     134590.000000  
mean       12696.123256  
std         6928.108124  
min         2600.000000  
25%         7600.000000  
50%        11700.000000  
75%        15300.000000  
max        45220.000000

In [10]:
df_hotels.shape

(25, 4)

In [11]:
df_hotels.head(3)

property_id  property_name  category    city
0        16558   Atliq Grands    Luxury   Delhi
1        16559  Atliq Exotica    Luxury  Mumbai
2        16560     Atliq City  Business   Delhi

In [12]:
df_hotels.category.value_counts()

Luxury      16
Business     9
Name: category, dtype: int64

In [13]:
# Count the number of hotels per city
hotel_counts = df_hotels['city'].value_counts().reset_index()
hotel_counts.columns = ['City', 'Number of Hotels']

# Create a horizontal bar chart
fig = px.bar(hotel_counts, 
             x='Number of Hotels', 
             y='City', 
             title='Number of Hotels per City',
             orientation='h',
             color='Number of Hotels',
             color_continuous_scale=px.colors.sequential.Peach)

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='Number of Hotels',
    yaxis_title='City',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
     autosize=False,width=950,height=500
)

# Adding value annotations on the bars
fig.for_each_trace(lambda t: t.update(text=[f"{v}" for v in t.x], textposition='outside'))

# Show the plot
fig.show()


#### Mumbai city is having more number of hotels

In [14]:
df_agg_bookings.head(3)

property_id check_in_date room_category  successful_bookings  capacity
0        16559     01-May-22           RT1                   25        30
1        19562     01-May-22           RT1                   28        30
2        19563     01-May-22           RT1                   23        30

In [15]:
df_agg_bookings.property_id.unique()

array([16559, 19562, 19563, 17558, 16558, 17560, 19558, 19560, 17561,
       16560, 16561, 16562, 16563, 17559, 17562, 17563, 18558, 18559,
       18561, 18562, 18563, 19559, 19561, 17564, 18560], dtype=int64)

In [16]:
df_agg_bookings.groupby("property_id")["successful_bookings"].sum()

property_id
16558    3153
16559    7338
16560    4693
16561    4418
16562    4820
16563    7147
17558    5036
17559    6142
17560    6013
17561    5183
17562    3424
17563    6337
17564    3982
18558    4475
18559    5256
18560    6638
18561    6458
18562    7333
18563    4728
19558    4371
19559    4705
19560    5979
19561    5736
19562    5812
19563    5413
Name: successful_bookings, dtype: int64

In [17]:
df_agg_bookings[df_agg_bookings.successful_bookings > df_agg_bookings.capacity]

Empty DataFrame
Columns: [property_id, check_in_date, room_category, successful_bookings, capacity]
Index: []

In [18]:
df_agg_bookings.capacity.max()

50

In [19]:
df_bookings.describe()

property_id      no_guests  ratings_given  revenue_generated  \
count  134590.000000  134590.000000   56683.000000      134590.000000   
mean    18061.113493       2.036808       3.619004       14916.013188   
std      1093.055847       1.031766       1.235009        6452.868072   
min     16558.000000       1.000000       1.000000        6500.000000   
25%     17558.000000       1.000000       3.000000        9900.000000   
50%     17564.000000       2.000000       4.000000       13500.000000   
75%     18563.000000       2.000000       5.000000       18000.000000   
max     19563.000000       6.000000       5.000000       45220.000000   

       revenue_realized  
count     134590.000000  
mean       12696.123256  
std         6928.108124  
min         2600.000000  
25%         7600.000000  
50%        11700.000000  
75%        15300.000000  
max        45220.000000

In [20]:
df_bookings[df_bookings.no_guests<=0]

Empty DataFrame
Columns: [booking_id, property_id, booking_date, check_in_date, checkout_date, no_guests, room_category, booking_platform, ratings_given, booking_status, revenue_generated, revenue_realized]
Index: []

In [21]:
df_bookings = df_bookings[df_bookings.no_guests>0]

In [22]:
df_bookings.shape

(134590, 12)

In [23]:
df_bookings.revenue_generated.min(), df_bookings.revenue_generated.max()

(6500, 45220)

In [24]:
df_bookings.revenue_generated.mean(), df_bookings.revenue_generated.median()

(14916.013188201203, 13500.0)

In [25]:
avg, std = df_bookings.revenue_generated.mean(), df_bookings.revenue_generated.std()

### Removing Outliers from the dataset

In [26]:
higher_limit = avg + 3*std
higher_limit

34274.617403506796

In [27]:
lower_limit = avg - 3*std
lower_limit

-4442.59102710439

In [28]:
df_bookings[df_bookings.revenue_generated<=0]

Empty DataFrame
Columns: [booking_id, property_id, booking_date, check_in_date, checkout_date, no_guests, room_category, booking_platform, ratings_given, booking_status, revenue_generated, revenue_realized]
Index: []

In [29]:
df_bookings[df_bookings.revenue_generated>higher_limit]

booking_id  property_id booking_date check_in_date  \
137      May012216559RT41        16559   2022-04-27    2022-05-01   
139      May012216559RT43        16559   2022-05-01    2022-05-01   
143      May012216559RT47        16559   2022-04-28    2022-05-01   
149     May012216559RT413        16559   2022-04-24    2022-05-01   
153     May012216559RT417        16559   2022-04-30    2022-05-01   
...                   ...          ...          ...           ...   
134467   Jul312219562RT45        19562   2022-07-28    2022-07-31   
134469   Jul312219562RT47        19562   2022-07-10    2022-07-31   
134474  Jul312219562RT412        19562   2022-07-25    2022-07-31   
134581   Jul312217564RT42        17564   2022-07-31    2022-07-31   
134586   Jul312217564RT47        17564   2022-07-30    2022-07-31   

       checkout_date  no_guests room_category booking_platform  ratings_given  \
137       2022-05-07          4           RT4           others            NaN   
139       2022-05-02          6           RT4         tripster            3.0   
143       2022-05-03          3           RT4           others            5.0   
149       2022-05-07          5           RT4          logtrip            NaN   
153       2022-05-02          4           RT4           others            NaN   
...              ...        ...           ...              ...            ...   
134467    2022-08-01          6           RT4     makeyourtrip            4.0   
134469    2022-08-06          5           RT4     makeyourtrip            NaN   
134474    2022-08-06          5           RT4   direct offline            5.0   
134581    2022-08-01          4           RT4     makeyourtrip            4.0   
134586    2022-08-01          4           RT4          logtrip            2.0   

       booking_status  revenue_generated  revenue_realized  
137       Checked Out              38760             38760  
139       Checked Out              45220             45220  
143       Checked Out              35530             35530  
149       Checked Out              41990             41990  
153         Cancelled              38760             15504  
...               ...                ...               ...  
134467    Checked Out              39900             39900  
134469      Cancelled              37050             14820  
134474    Checked Out              37050             37050  
134581    Checked Out              38760             38760  
134586    Checked Out              38760             38760  

[1520 rows x 12 columns]

In [30]:
df_bookings = df_bookings[df_bookings.revenue_generated<=higher_limit]
df_bookings.shape

(133070, 12)

In [31]:
df_bookings.revenue_realized.describe().round(2)

count    133070.00
mean      12468.78
std        6537.75
min        2600.00
25%        7600.00
50%       11400.00
75%       15300.00
max       34200.00
Name: revenue_realized, dtype: float64

In [32]:
higher_limit = df_bookings.revenue_realized.mean() + 3*df_bookings.revenue_realized.std()
higher_limit

32082.021279982062

In [33]:
df_bookings[df_bookings.revenue_realized>higher_limit]

booking_id  property_id booking_date check_in_date  \
140      May012216559RT44        16559   2022-04-28    2022-05-01   
141      May012216559RT45        16559   2022-04-28    2022-05-01   
144      May012216559RT48        16559   2022-04-26    2022-05-01   
145      May012216559RT49        16559   2022-04-28    2022-05-01   
146     May012216559RT410        16559   2022-04-30    2022-05-01   
...                   ...          ...          ...           ...   
134580   Jul312217564RT41        17564   2022-07-31    2022-07-31   
134582   Jul312217564RT43        17564   2022-07-31    2022-07-31   
134584   Jul312217564RT45        17564   2022-07-30    2022-07-31   
134585   Jul312217564RT46        17564   2022-07-29    2022-07-31   
134588   Jul312217564RT49        17564   2022-07-29    2022-07-31   

       checkout_date  no_guests room_category booking_platform  ratings_given  \
140       2022-05-05          2           RT4           others            NaN   
141       2022-05-04          2           RT4           others            5.0   
144       2022-05-02          2           RT4     makeyourtrip            5.0   
145       2022-05-03          2           RT4           others            NaN   
146       2022-05-06          2           RT4          logtrip            NaN   
...              ...        ...           ...              ...            ...   
134580    2022-08-02          1           RT4    direct online            NaN   
134582    2022-08-03          1           RT4           others            1.0   
134584    2022-08-01          2           RT4           others            2.0   
134585    2022-08-03          1           RT4     makeyourtrip            2.0   
134588    2022-08-01          2           RT4          logtrip            2.0   

       booking_status  revenue_generated  revenue_realized  
140       Checked Out              32300             32300  
141       Checked Out              32300             32300  
144       Checked Out              32300             32300  
145       Checked Out              32300             32300  
146           No Show              32300             32300  
...               ...                ...               ...  
134580    Checked Out              32300             32300  
134582    Checked Out              32300             32300  
134584    Checked Out              32300             32300  
134585    Checked Out              32300             32300  
134588    Checked Out              32300             32300  

[3256 rows x 12 columns]

In [34]:
df_bookings[df_bookings.room_category=="RT4"].revenue_realized.describe().round(2)

count    14553.00
mean     22483.44
std       8338.25
min       7600.00
25%      19000.00
50%      22800.00
75%      31350.00
max      34200.00
Name: revenue_realized, dtype: float64

In [35]:
df_bookings[df_bookings.booking_id=="May012216558RT213"]

booking_id  property_id booking_date check_in_date checkout_date  \
30  May012216558RT213        16558   2022-04-29    2022-05-01    2022-05-02   

    no_guests room_category booking_platform  ratings_given booking_status  \
30          2           RT2          logtrip            4.0    Checked Out   

    revenue_generated  revenue_realized  
30              12600             12600

In [36]:
df_bookings.isnull().sum()

booking_id               0
property_id              0
booking_date             0
check_in_date            0
checkout_date            0
no_guests                0
room_category            0
booking_platform         0
ratings_given        77054
booking_status           0
revenue_generated        0
revenue_realized         0
dtype: int64

In [37]:
df_agg_bookings.isnull().sum()

property_id            0
check_in_date          0
room_category          0
successful_bookings    0
capacity               0
dtype: int64

In [38]:
df_agg_bookings[df_agg_bookings.capacity.isna()]

Empty DataFrame
Columns: [property_id, check_in_date, room_category, successful_bookings, capacity]
Index: []

In [39]:
df_agg_bookings.capacity.median()

25.0

In [40]:
df_agg_bookings.capacity.fillna(df_agg_bookings.capacity.median(), inplace=True)

In [41]:
df_agg_bookings.loc[[8,15]]

property_id check_in_date room_category  successful_bookings  capacity
8         17561     01-May-22           RT1                   22        26
15        17563     01-May-22           RT1                   21        25

In [42]:
df_agg_bookings[df_agg_bookings.successful_bookings>df_agg_bookings.capacity]


Empty DataFrame
Columns: [property_id, check_in_date, room_category, successful_bookings, capacity]
Index: []

In [43]:
df_agg_bookings.head(3)

property_id check_in_date room_category  successful_bookings  capacity
0        16559     01-May-22           RT1                   25        30
1        19562     01-May-22           RT1                   28        30
2        19563     01-May-22           RT1                   23        30

In [44]:
df_agg_bookings['occ_pct'] = df_agg_bookings.apply(lambda row: row['successful_bookings']/row['capacity'], axis=1)

In [45]:
new_col = df_agg_bookings.apply(lambda row: row['successful_bookings']/row['capacity'], axis=1)
df_agg_bookings = df_agg_bookings.assign(occ_pct=new_col.values)
df_agg_bookings.head(3)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        19562     01-May-22           RT1                   28        30   
2        19563     01-May-22           RT1                   23        30   

    occ_pct  
0  0.833333  
1  0.933333  
2  0.766667

In [46]:
df_agg_bookings['occ_pct'] = df_agg_bookings['occ_pct'].apply(lambda x: round(x*100, 2))
df_agg_bookings.head(3)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        19562     01-May-22           RT1                   28        30   
2        19563     01-May-22           RT1                   23        30   

   occ_pct  
0    83.33  
1    93.33  
2    76.67

#### Occupany percentage is calculated

In [47]:
df_bookings.head()


booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   
3  May012216558RT14        16558   2022-04-28    2022-05-01    2022-05-02   
4  May012216558RT15        16558   2022-04-27    2022-05-01    2022-05-02   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   
3          2           RT1           others            NaN      Cancelled   
4          4           RT1    direct online            5.0    Checked Out   

   revenue_generated  revenue_realized  
0              10010             10010  
1               9100              3640  
2               9100              9100  
3               9100              3640  
4              10920             10920

In [48]:
df_agg_bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9200 entries, 0 to 9199
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   property_id          9200 non-null   int64  
 1   check_in_date        9200 non-null   object 
 2   room_category        9200 non-null   object 
 3   successful_bookings  9200 non-null   int64  
 4   capacity             9200 non-null   int64  
 5   occ_pct              9200 non-null   float64
dtypes: float64(1), int64(3), object(2)
memory usage: 431.4+ KB


In [49]:
df_agg_bookings.head(3)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        19562     01-May-22           RT1                   28        30   
2        19563     01-May-22           RT1                   23        30   

   occ_pct  
0    83.33  
1    93.33  
2    76.67

In [50]:
avg_room_category_occupancy_rate=df_agg_bookings.groupby("room_category")["occ_pct"].mean().round(2)
avg_room_category_occupancy_rate

room_category
RT1    57.92
RT2    58.01
RT3    58.03
RT4    59.28
Name: occ_pct, dtype: float64

In [51]:
# Create a DataFrame from the Series
avg_room_occupancy_df = avg_room_category_occupancy_rate.reset_index()
avg_room_occupancy_df.columns = ['room_category', 'Average Occupancy Rate']

# Create a bar chart
fig = px.bar(avg_room_occupancy_df, 
              x='room_category', 
              y='Average Occupancy Rate', 
              title='Average Occupancy Rate per Room Category',
              color='Average Occupancy Rate',
              color_continuous_scale=px.colors.sequential.Viridis)

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='Room Category',
    yaxis_title='Average Occupancy Rate (%)',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
     autosize=False,width=950,height=550
)

# Adding value annotations on the bars
fig.for_each_trace(lambda t: t.update(text=[f"{v}%" for v in t.y], textposition='outside'))

# Show the plot
fig.show()

#### RT4 room category occupancy percentage is more compared to other categories

In [52]:
df = pd.merge(df_agg_bookings, df_rooms, left_on="room_category", right_on="room_id")
df.head(4)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        19562     01-May-22           RT1                   28        30   
2        19563     01-May-22           RT1                   23        30   
3        17558     01-May-22           RT1                   13        19   

   occ_pct room_id room_class  
0    83.33     RT1   Standard  
1    93.33     RT1   Standard  
2    76.67     RT1   Standard  
3    68.42     RT1   Standard

In [53]:
df.drop("room_id",axis=1, inplace=True)
df.head(4)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        19562     01-May-22           RT1                   28        30   
2        19563     01-May-22           RT1                   23        30   
3        17558     01-May-22           RT1                   13        19   

   occ_pct room_class  
0    83.33   Standard  
1    93.33   Standard  
2    76.67   Standard  
3    68.42   Standard

In [54]:
avg_room_class_occupancy_rate=df.groupby("room_class")["occ_pct"].mean().round(2)
avg_room_class_occupancy_rate

room_class
Elite           58.01
Premium         58.03
Presidential    59.28
Standard        57.92
Name: occ_pct, dtype: float64

In [55]:
# Create a DataFrame from the Series
avg_room_class_occupancy_df = avg_room_class_occupancy_rate.reset_index()
avg_room_class_occupancy_df.columns = ['room_class', 'Average Occupancy Rate']

# Create a bar chart
fig = px.bar(avg_room_class_occupancy_df, 
              x='room_class', 
              y='Average Occupancy Rate', 
              title='Average Occupancy Rate per Room Class',
              color='Average Occupancy Rate',
              color_continuous_scale=px.colors.sequential.Agsunset)

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='Room Class',
    yaxis_title='Average Occupancy Rate (%)',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
     autosize=False,width=950,height=550
)

# Adding value annotations on the bars
fig.for_each_trace(lambda t: t.update(text=[f"{v}%" for v in t.y], textposition='outside'))

# Show the plot
fig.show()

#### Presidential room class occupancy is the highest among other room classes

In [56]:
df[df.room_class=="Standard"].occ_pct.mean()

57.919021739130486

In [57]:
df_hotels.head(3)

property_id  property_name  category    city
0        16558   Atliq Grands    Luxury   Delhi
1        16559  Atliq Exotica    Luxury  Mumbai
2        16560     Atliq City  Business   Delhi

In [58]:
df = pd.merge(df, df_hotels, on="property_id")
df.head(3)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        16559     02-May-22           RT1                   20        30   
2        16559     03-May-22           RT1                   17        30   

   occ_pct room_class  property_name category    city  
0    83.33   Standard  Atliq Exotica   Luxury  Mumbai  
1    66.67   Standard  Atliq Exotica   Luxury  Mumbai  
2    56.67   Standard  Atliq Exotica   Luxury  Mumbai

In [59]:
avg_occupancy_rate_per_city = df.groupby("city")["occ_pct"].mean().round(2)
avg_occupancy_rate_per_city

city
Bangalore    56.34
Delhi        61.52
Hyderabad    58.12
Mumbai       57.92
Name: occ_pct, dtype: float64

In [60]:
# Create a DataFrame from the Series
avg_occupancy_df = avg_occupancy_rate_per_city.reset_index()
avg_occupancy_df.columns = ['City', 'Average Occupancy Rate']

# Create a bar chart
fig = px.bar(avg_occupancy_df, 
              x='City', 
              y='Average Occupancy Rate', 
              title='Average Occupancy Rate per City',
              color='Average Occupancy Rate',
              color_continuous_scale=px.colors.sequential.Plasma)

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='City',
    yaxis_title='Average Occupancy Rate (%)',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
     autosize=False,width=950,height=550
)

# Adding value annotations on the bars
fig.for_each_trace(lambda t: t.update(text=[f"{v}%" for v in t.y], textposition='outside'))

# Show the plot
fig.show()



#### Delhi is leading in the rooms occupancy percentage

In [61]:
df_date.head(3)

date  mmm yy week no  day_type
0  01-May-22  May 22    W 19   weekend
1  02-May-22  May 22    W 19  weekeday
2  03-May-22  May 22    W 19  weekeday

In [62]:
df = pd.merge(df, df_date, left_on="check_in_date", right_on="date")
df.head(3)

property_id check_in_date room_category  successful_bookings  capacity  \
0        16559     01-May-22           RT1                   25        30   
1        16559     01-May-22           RT2                   35        41   
2        16559     01-May-22           RT3                   27        32   

   occ_pct room_class  property_name category    city       date  mmm yy  \
0    83.33   Standard  Atliq Exotica   Luxury  Mumbai  01-May-22  May 22   
1    85.37      Elite  Atliq Exotica   Luxury  Mumbai  01-May-22  May 22   
2    84.38    Premium  Atliq Exotica   Luxury  Mumbai  01-May-22  May 22   

  week no day_type  
0    W 19  weekend  
1    W 19  weekend  
2    W 19  weekend

In [63]:
df.groupby("day_type")["occ_pct"].mean().round(2)

day_type
weekeday    51.81
weekend     73.96
Name: occ_pct, dtype: float64

In [64]:
# Calculate the mean occupancy percentage by day type
mean_occ_pct = df.groupby("day_type")["occ_pct"].mean().round(2).reset_index()

# Create a pie chart
fig = px.pie(mean_occ_pct, 
             names='day_type', 
             values='occ_pct', 
             title='Occupancy Percentage by Day Type',
             hole=0.3,  # Optional: make it a donut chart
             color='day_type',  # Use colors based on day type
             color_discrete_sequence=px.colors.sequential.RdBu)

# Updating layout for better appearance
fig.update_traces(textinfo='percent+label', pull=[0.05] * len(mean_occ_pct))  # Slightly pull out each slice
fig.update_layout(
    title_font=dict(size=16, family='Arial', color='black'),
    autosize=False,width=800,height=800
)

# Show the plot
fig.show()


#### Rooms are booked more for weekends compared to weekdays

In [65]:
df_june_22 = df[df["mmm yy"]=="Jun 22"]
df_june_22.head(4)

property_id check_in_date room_category  successful_bookings  capacity  \
3100        16559     01-Jun-22           RT1                   14        30   
3101        16559     01-Jun-22           RT2                   21        41   
3102        16559     01-Jun-22           RT3                   15        32   
3103        16559     01-Jun-22           RT4                    8        18   

      occ_pct    room_class  property_name category    city       date  \
3100    46.67      Standard  Atliq Exotica   Luxury  Mumbai  01-Jun-22   
3101    51.22         Elite  Atliq Exotica   Luxury  Mumbai  01-Jun-22   
3102    46.88       Premium  Atliq Exotica   Luxury  Mumbai  01-Jun-22   
3103    44.44  Presidential  Atliq Exotica   Luxury  Mumbai  01-Jun-22   

      mmm yy week no  day_type  
3100  Jun 22    W 23  weekeday  
3101  Jun 22    W 23  weekeday  
3102  Jun 22    W 23  weekeday  
3103  Jun 22    W 23  weekeday

In [66]:
df_june_22.groupby('city')['occ_pct'].mean().round(2).sort_values(ascending=False)

city
Delhi        61.46
Mumbai       57.79
Hyderabad    57.69
Bangalore    55.85
Name: occ_pct, dtype: float64

In [67]:
df_august = pd.read_csv("datasets/new_data_august.csv")
df_august.head(3)

property_id  property_name  category       city room_category room_class  \
0        16559  Atliq Exotica    Luxury     Mumbai           RT1   Standard   
1        19562      Atliq Bay    Luxury  Bangalore           RT1   Standard   
2        19563   Atliq Palace  Business  Bangalore           RT1   Standard   

  check_in_date  mmm yy week no  day_type  successful_bookings  capacity  \
0     01-Aug-22  Aug-22    W 32  weekeday                   30        30   
1     01-Aug-22  Aug-22    W 32  weekeday                   21        30   
2     01-Aug-22  Aug-22    W 32  weekeday                   23        30   

     occ%  
0  100.00  
1   70.00  
2   76.67

In [68]:
df_bookings.head()

booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   
3  May012216558RT14        16558   2022-04-28    2022-05-01    2022-05-02   
4  May012216558RT15        16558   2022-04-27    2022-05-01    2022-05-02   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   
3          2           RT1           others            NaN      Cancelled   
4          4           RT1    direct online            5.0    Checked Out   

   revenue_generated  revenue_realized  
0              10010             10010  
1               9100              3640  
2               9100              9100  
3               9100              3640  
4              10920             10920

In [69]:
df_august.shape

(7, 13)

In [70]:
df_merge = pd.concat([df, df_august], ignore_index = True, axis = 0)
df_merge.tail(5)

property_id check_in_date room_category  successful_bookings  capacity  \
9202        19563     01-Aug-22           RT1                   23        30   
9203        19558     01-Aug-22           RT1                   30        40   
9204        19560     01-Aug-22           RT1                   20        26   
9205        17561     01-Aug-22           RT1                   18        26   
9206        17564     01-Aug-22           RT1                   10        16   

      occ_pct room_class  property_name  category       city date  mmm yy  \
9202      NaN   Standard   Atliq Palace  Business  Bangalore  NaN  Aug-22   
9203      NaN   Standard   Atliq Grands    Luxury  Bangalore  NaN  Aug-22   
9204      NaN   Standard     Atliq City  Business  Bangalore  NaN  Aug-22   
9205      NaN   Standard      Atliq Blu    Luxury     Mumbai  NaN  Aug-22   
9206      NaN   Standard  Atliq Seasons  Business     Mumbai  NaN  Aug-22   

     week no  day_type   occ%  
9202    W 32  weekeday  76.67  
9203    W 32  weekeday  75.00  
9204    W 32  weekeday  76.92  
9205    W 32  weekeday  69.23  
9206    W 32  weekeday  62.50

In [71]:
df_merge.shape

(9207, 15)

In [72]:
df_hotels.head(3)

property_id  property_name  category    city
0        16558   Atliq Grands    Luxury   Delhi
1        16559  Atliq Exotica    Luxury  Mumbai
2        16560     Atliq City  Business   Delhi

In [73]:
df_bookings_all = pd.merge(df_bookings, df_hotels, on="property_id")
df_bookings_all.head(3)

booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   

   revenue_generated  revenue_realized property_name category   city  
0              10010             10010  Atliq Grands   Luxury  Delhi  
1               9100              3640  Atliq Grands   Luxury  Delhi  
2               9100              9100  Atliq Grands   Luxury  Delhi

In [74]:
df_bookings_all.groupby("city")["revenue_realized"].sum()

city
Bangalore    415122840
Delhi        289472918
Hyderabad    325232870
Mumbai       629391323
Name: revenue_realized, dtype: int64

In [75]:
# Calculate total revenue by city
city_revenue = df_bookings_all.groupby("city")["revenue_realized"].sum()

# Convert revenue to millions
city_revenue_millions = city_revenue / 1_000_000

# Create a DataFrame for Plotly
city_revenue_df = city_revenue_millions.reset_index()
city_revenue_df.columns = ['City', 'Total Revenue (in Millions)']

# Create a horizontal bar chart
fig = px.bar(city_revenue_df, 
              x='Total Revenue (in Millions)', 
              y='City', 
              title='Total Revenue Realized by City (in Millions)',
              orientation='h',
              color='Total Revenue (in Millions)',
              color_continuous_scale=px.colors.sequential.Teal)

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='Total Revenue (in Millions)',
    yaxis_title='City',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
    autosize=False,width=950,height=500
)

# Adding value annotations on the bars
fig.for_each_trace(lambda t: t.update(text=[f'{v:,.0f}M' for v in t.x], textposition='inside'))

# Show the plot
fig.show()


#### Out of all cities Mumbai generated more revenue

In [76]:
df_date["mmm yy"].unique()

array(['May 22', 'Jun 22', 'Jul 22'], dtype=object)

In [77]:
df_bookings_all.head(3)

booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   

   revenue_generated  revenue_realized property_name category   city  
0              10010             10010  Atliq Grands   Luxury  Delhi  
1               9100              3640  Atliq Grands   Luxury  Delhi  
2               9100              9100  Atliq Grands   Luxury  Delhi

In [78]:
df_date.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92 entries, 0 to 91
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   date      92 non-null     object
 1   mmm yy    92 non-null     object
 2   week no   92 non-null     object
 3   day_type  92 non-null     object
dtypes: object(4)
memory usage: 3.0+ KB


In [79]:
df_date["date"] = pd.to_datetime(df_date["date"], errors='coerce')
df_date.head()

date  mmm yy week no  day_type
0 2022-05-01  May 22    W 19   weekend
1 2022-05-02  May 22    W 19  weekeday
2 2022-05-03  May 22    W 19  weekeday
3 2022-05-04  May 22    W 19  weekeday
4 2022-05-05  May 22    W 19  weekeday

In [80]:
df_bookings_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 133070 entries, 0 to 133069
Data columns (total 15 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   booking_id         133070 non-null  object 
 1   property_id        133070 non-null  int64  
 2   booking_date       133070 non-null  object 
 3   check_in_date      133070 non-null  object 
 4   checkout_date      133070 non-null  object 
 5   no_guests          133070 non-null  int64  
 6   room_category      133070 non-null  object 
 7   booking_platform   133070 non-null  object 
 8   ratings_given      56016 non-null   float64
 9   booking_status     133070 non-null  object 
 10  revenue_generated  133070 non-null  int64  
 11  revenue_realized   133070 non-null  int64  
 12  property_name      133070 non-null  object 
 13  category           133070 non-null  object 
 14  city               133070 non-null  object 
dtypes: float64(1), int64(4), object(10)
memory usage: 1

In [81]:
df_bookings_all["check_in_date"] = pd.to_datetime(df_bookings_all["check_in_date"], errors='coerce')
df_bookings_all.head(3)

booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   

   revenue_generated  revenue_realized property_name category   city  
0              10010             10010  Atliq Grands   Luxury  Delhi  
1               9100              3640  Atliq Grands   Luxury  Delhi  
2               9100              9100  Atliq Grands   Luxury  Delhi

In [82]:
df_bookings_all = pd.merge(df_bookings_all, df_date, left_on="check_in_date", right_on="date")
df_bookings_all.head(3)

booking_id  property_id booking_date check_in_date checkout_date  \
0  May012216558RT11        16558   2022-04-27    2022-05-01    2022-05-02   
1  May012216558RT12        16558   2022-04-30    2022-05-01    2022-05-02   
2  May012216558RT13        16558   2022-04-28    2022-05-01    2022-05-04   

   no_guests room_category booking_platform  ratings_given booking_status  \
0          3           RT1    direct online            1.0    Checked Out   
1          2           RT1           others            NaN      Cancelled   
2          2           RT1          logtrip            5.0    Checked Out   

   revenue_generated  revenue_realized property_name category   city  \
0              10010             10010  Atliq Grands   Luxury  Delhi   
1               9100              3640  Atliq Grands   Luxury  Delhi   
2               9100              9100  Atliq Grands   Luxury  Delhi   

        date  mmm yy week no day_type  
0 2022-05-01  May 22    W 19  weekend  
1 2022-05-01  May 22    W 19  weekend  
2 2022-05-01  May 22    W 19  weekend

In [83]:
df_bookings_all.groupby("mmm yy")["revenue_realized"].sum().round(2)

mmm yy
Jul 22    556460174
Jun 22    537219271
May 22    565540506
Name: revenue_realized, dtype: int64

#### In month of May highest revenue is generated

In [84]:
df_bookings_all.property_name.unique()

array(['Atliq Grands', 'Atliq Exotica', 'Atliq City', 'Atliq Blu',
       'Atliq Bay', 'Atliq Palace', 'Atliq Seasons'], dtype=object)

In [85]:
df_bookings_all.groupby("property_name")["revenue_realized"].sum().round(2).sort_values()

property_name
Atliq Seasons     62430375
Atliq Grands     207800708
Atliq Bay        255735290
Atliq Blu        256643260
Atliq City       279554213
Atliq Palace     294193161
Atliq Exotica    302862944
Name: revenue_realized, dtype: int64

In [86]:
# Calculate total revenue by hotel
revenue_by_hotel = df_bookings_all.groupby("property_name")["revenue_realized"].sum().round(2).sort_values()
revenue_by_hotel_millions = revenue_by_hotel / 1_000_000

# Create a DataFrame for Plotly
revenue_df = revenue_by_hotel_millions.reset_index()
revenue_df.columns = ['Hotel', 'Total Revenue (in Millions)']

# Create a vertical bar chart
fig = px.bar(revenue_df, 
              x='Hotel', 
              y='Total Revenue (in Millions)', 
              title='Total Revenue Realized by Hotel (in Millions)',
              color='Total Revenue (in Millions)',
              color_continuous_scale=px.colors.sequential.Sunsetdark   
            )

# Updating layout for better appearance
fig.update_layout(
    xaxis_title='Hotel',
    yaxis_title='Total Revenue(in Millions)',
    font=dict(size=14),
    title_font=dict(size=16, family='Arial', color='black'),
    xaxis_tickangle=-45,  # Rotate x-axis labels for better readability
    autosize=False,width=950,height=600
)

# Adding value annotations on the bars
fig.for_each_trace(lambda t: t.update(text=[f'{v:.2f}M' for v in t.y], textposition='outside'))

# Show the plot
fig.show()


#### Atliq Exotica generated more revenue than other hotels

In [87]:
df_bookings_all.groupby("city")["ratings_given"].mean().round(2)

city
Bangalore    3.41
Delhi        3.78
Hyderabad    3.66
Mumbai       3.65
Name: ratings_given, dtype: float64

#### Customer gave the highest rating to Delhi

In [88]:
# Calculate total revenue by booking platform
revenue_by_platform = df_bookings_all.groupby("booking_platform")["revenue_realized"].sum().reset_index()

# Create a pie chart
fig = px.pie(revenue_by_platform, 
             names='booking_platform', 
             values='revenue_realized', 
             title='Revenue Realized by Booking Platform',
             hole=0.3,  # Optional: make it a donut chart
             color='booking_platform',  # Color based on platform
             color_discrete_sequence=px.colors.qualitative.Antique)

# Updating layout for better appearance
fig.update_traces(textinfo='percent+label')  # Show both percentage and label
fig.update_layout(
    title_font=dict(size=18, family='Arial', color='black'), autosize=False,width=800,height=800
)

# Show the plot
fig.show()


#### Apart fro other platforms the second highest revenue is generated by makeyoutrip platfom

# Insights

### Average Occupancy by Room Type
 * Presidential rooms have the highest average occupancy (59.28%).
 * Elite and Premium rooms have the lowest average occupancy (58.01% and 58.03%)

### Average Occupancy Rate per City
 * Delhi has the highest average occupancy rate (61.52%).
 * Bangalore has the lowest average occupancy rate (56.34%)

### Weekday vs. Weekend Occupancy
* Occupancy is significantly higher on weekends (73.96%) compared to weekdays (51.81%)

### Occupancy Rates in June
 * Delhi had the highest occupancy in June (61.46%).
 * Bangalore had the lowest occupancy in June (55.85%)
 
### Revenue Realized per City
 * Mumbai generates the highest revenue (USD 629M)
 * Bangalore generates the lowest revenue (USD 415M)

### Revenue by Hotel Type
 * Atliq Exotica has the highest revenue (USD 302.86M)
 * Atliq Seasons has the lowest revenue (USD 62.43M)

### Average Rating per City
 * Delhi has the highest average rating (3.78)
 * Bangalore has the lowest average rating (3.41)

# Suggestions

### 1. Promote Less Popular Room Types and Cities
 * Use targeted promotions and partnerships to boost occupancy in less popular room types and cities

### 2. Enhance Weekday Occupancy
 * Create attractive weekday offers and corporate partnerships to balance occupancy rates

### 3. Adjust Strategies Based on Revenue and Ratings
 * Focus on cities and hotel types with lower revenue and ratings for targeted improvements

### 4. Utilize Data for Decision-Making
 * Regularly integrate and analyze new data to make informed decisions and forecasts

### 5. Improve Customer Satisfaction
 * Invest in training and resources to enhance guest experience, particularly in cities and hotel types with lower average    ratings

### 6. Optimize Marketing Efforts
 * Tailor marketing campaigns based on historical data to effectively target high-potential markets and improve overall        occupancy and revenue.